pkg/ruler/manager.go

```diff
@@ -105,9 +105,17 @@ func (r *DefaultMultiTenantManager) SyncRuleGroups(ctx context.Context, ruleGrou
 		RemoveFederatedRuleGroups(ruleGroups)
 	}
 
+	// Sync the rules to disk and then update the user's Prometheus Rules Manager.
+	// Since users are different, we can sync rules in parallel.
+	var wg sync.WaitGroup
 	for userID, ruleGroup := range ruleGroups {
-		r.syncRulesToManager(ctx, userID, ruleGroup)
+		wg.Add(1)
+		go func(id string, rg rulespb.RuleGroupList) {
```

Comment:
I think we should limit concurrency (e.g. 10). In dskit "concurrent" package there's an utility you could use to limit concurrency.

Reply:
Thank you, the PR is updated.